Adding an NPM guides section (#9394)
DataDog/documentation: content/en/network_performance_monitoring/guide/manage_traffic_costs_with_npm.md

---
title: Manage Cloud Traffic Costs with NPM
kind: guide
---

> Traffic is expensive, especially in the cloud. Cloud providers charge different prices for traffic, depending on whether it is flowing within an availability zone, between availability zones (AZs), between particular regions, or to the open internet. Cross-regional and egress traffic is not only the most expensive, but also the most vulnerable to errors, latency, and security threats. 

Reviewer: ```suggestion
Traffic is expensive, especially in the cloud. Cloud providers charge different prices for traffic, depending on whether it is flowing within an availability zone (AZ), between AZs, between particular regions, or to the open internet. Cross-regional and egress traffic is not only the most expensive, but also the most vulnerable to errors, latency, and security threats. 
```

> Network Performance Monitoring allows you to track all of the traffic patterns described above by mapping dependencies between any tags in Datadog, including service, container, availability zone, region, datacenter, etc… This insight into your dependencies and the amount of traffic volume they produce (which is ultimately what cloud providers charge for) can be used to monitor and optimize on your traffic-related costs. 

Reviewer: ```suggestion
Network Performance Monitoring (NPM) allows you to track all of the traffic patterns described above by mapping dependencies between any tags in Datadog, including service, container, availability zone, region, datacenter, etc. This insight into your dependencies and the traffic volume they produce (which is ultimately what cloud providers charge for) can be used to monitor and optimize your traffic-related costs. 
```

## Datadog’s story

> When we began our Kubernetes migration, migrating stateless services was (expectedly) much faster and easier than migrating our stateful services (e.g. Kafka), so we migrated our stateless services first. The outcome was an terabytes of new cross-AZ traffic between the stateful services (all in one AZ) and stateless services (spread across the other AZs), which led our cloud bill to increase drastically and unexpectedly. We used our own NPM product to identify the root cause - a suboptimal migration strategy and consequently, inefficient and costly network communication. By sharding our stateful services, we were ultimately able to see significant (million-dollar) reductions in our cloud provider traffic costs.

Reviewer: ```suggestion
When Datadog migrated to Kubernetes, migrating stateless services was (expectedly) much faster and easier than migrating stateful services (e.g. Kafka), stateless services were first. The outcome was a terabytes of new cross-AZ traffic between the stateful services (all in one AZ) and stateless services (spread across the other AZs), which led the cloud bill to increase drastically and unexpectedly. Datadog used its own NPM product to identify the root cause: a suboptimal migration strategy and consequently, inefficient and costly network communication. Sharding stateful services ultimately led to significant reductions in our cloud provider traffic costs.
```

## Steps for managing traffic costs 

1. To find similar issues in your own environment, you can start by scoping your view to traffic between regions, 
    availability zones,
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/availability_zone.png" alt="Group flows by availability">}}
    and datacenters:
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/datacenter.png" alt="Group flows by datacenter">}}

>     An increase in your traffic bill is almost always tied to an increase in one of these types of traffic. In many cases you may want to group traffic by asymmetric search terms - meaning, you want to see the source of the traffic in terms of one tag, but want to see the destination in terms of another. This kind of asymmetric query can be used to identify costly dependencies between your on-prem datacenters and cloud regions, 

Reviewer: ```suggestion
    An increase in your traffic bill is almost always tied to an increase in one of these types of traffic. In many cases you may want to group traffic by asymmetric search terms. That is, you want to see the source of the traffic in terms of one tag, and also see the destination in terms of another tag. You can use this kind of asymmetric query to identify costly dependencies between your on-prem datacenters and cloud regions, 
```

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/aws_account.png" alt="Identify dependencies between datacenters and cloud regions">}}
    as well as between clouds. One particularly helpful view is to group the source of the traffic by service, and the destination across availability zones.

> 2. From here, isolate services that have the most traffic volume across multiple AZs. You can use filters inside the search bars to narrow down your query - so, for example, you can show only services that originate inside of one availability zone, and are sending traffic to any availability zone other than that one.

Reviewer: ```suggestion
2. From here, isolate services that have the most traffic volume across multiple AZs. You can use filters inside the search bars to narrow down your query. For example, you can show only services that originate inside of one availability zone, and are sending traffic to any other availability zone.
```

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/service_availability_zone.png" alt="Highlight any services that communicated across AZs">}}

>     The query above highlights any services that communicate across AZs, from us-east4-a to anywhere else. Since the table is already sorted by volume, the first few rows surface the chattiest services contributing to the most cross-AZ traffic. If we want to inspect the cross-infrastructure effects of one of these culprits, we can filter the Source to that particular service and view its traffic to every other availability zone.

Reviewer: ```suggestion
    The query above highlights any services that communicate from `us-east4-a` to anywhere else. Since the table is already sorted by volume, the first few rows surface the chattiest services contributing to the most cross-AZ traffic. If you want to inspect the cross-infrastructure effects of one of these culprits, you can filter the **Source** to that particular service and view its traffic to every other availability zone.
```

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/single_service.png" alt="Search for a single service.">}}

3. Similarly, you can use the team tag to identify engineering teams that generate the most, for instance, cross-regional traffic 
{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/team_region.png" alt="Use the team tag.">}}
or monitor your own team’s output specifically.
{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/region_region.png" alt="Use the region tag.">}}

> 4. To monitor costs from external traffic, scope your destination endpoints to public IPs using the IP Type facet.

Reviewer: ```suggestion
4. To monitor costs from external traffic, scope your destination endpoints to public IPs using the **IP Type** facet.
```

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/scope_destination_points.png" alt="Use the type facet.">}}

>     Then group your destination by `domain` to break down external traffic volume by where it’s going. Although you cannot install an agent on public servers, we are able to resolve IPs representing external and cloud endpoints to human-readable domain names. 

Reviewer: ```suggestion
    Then group your destination by `domain` to break down external traffic volume by where it is going. Although you cannot install a Datadog Agent on public servers, Datadog can resolve IPs representing external and cloud endpoints to human-readable domain names. 
```

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/dns_resolution.png" alt="Group by DNS.">}}

>     The example query above filters for traffic to AWS S3, elastic load balancers, APIs, and external ‘.com’ domains using substring wildcard entries (e.g. dns:*s3*).  

Reviewer: ```suggestion
    The example query above filters for traffic to AWS S3, elastic load balancers, APIs, and external `.com` domains using substring wildcard entries (for example: `dns:*s3*`).  
```

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/wildcard.png" alt="Search with wildcards">}}

> ## Visualizing Traffic Costs 

Reviewer: ```suggestion
## Visualizing traffic costs 
```

> You can visualize cross-AZ or inter-AZ traffic using the Network Map to quickly pinpoint bottlenecks. At Datadog, we also use this view to validate that our EU and US availability zones are not communicating, to ensure compliance with GDPR and protection of our customers' data. 

Reviewer: ```suggestion
You can visualize cross-AZ or inter-AZ traffic using the Network Map to quickly pinpoint bottlenecks. At Datadog,  this view is used to validate that the EU and US availability zones are not communicating, to ensure compliance with GDPR and protection of customers' data. 
```

> (Cross-AZ traffic)

Reviewer: ```suggestion
Cross-AZ traffic:
```

{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/cross-az-traffic.png" alt="Cross-AZ traffic">}}

> (Inter-AZ service-to-service traffic) 

Reviewer: ```suggestion
Inter-AZ service-to-service traffic:
```

{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/inter-az-service-to-service-traffic.png" alt="Inter-AZ service-to-service traffic">}}
Thick edges between the map's nodes, representing availability zones, indicates high volume of traffic flowing between them, which is what contributes to your costs.

> You can edit your preferences using the ‘Filter traffic’ button. In larger environments, we recommend scoping to just the most significant traffic sources by moving the sliders to include just the highest-volume dependencies.

Reviewer: ```suggestion
You can edit your preferences using the **Filter traffic** button. In larger environments, Datadog recommends scoping to just the most significant traffic sources by moving the sliders to include only the highest-volume dependencies.
```

{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/filter-traffic.png" alt="Scope your traffic">}}

> ## Graphing Traffic Costs 

Reviewer: ```suggestion
## Graphing traffic costs 
```

> We recommend tracking traffic volume metrics over time in dashboards and notebooks. You can graph traffic between any two endpoints using the same queries you would make in the network page. Just create a Timeseries Widget and select the ‘Network Traffic’ source from the dropdown menu.  

Reviewer: ```suggestion
Datadog recommends tracking traffic volume metrics over time in dashboards and notebooks. You can graph traffic between any two endpoints using the same queries you would make on the Network page. Create a **Timeseries Widget** and select the **Network Traffic** source from the dropdown menu.  
```

---

> title: Manage Traffic Costs with NPM

Reviewer 2: ```suggestion
title: Manage Cloud Traffic Costs with NPM
```

kind: guide
---

> Traffic is expensive, especially in the cloud. Cloud providers charge differently for traffic within an availability zone, traffic between availability zones (AZs), and traffic between regions (this pricing may differ by region). Cross-regional traffic is not only the most expensive, but also the most vulnerable to errors, latency, and security threats. 

Reviewer 2: ```suggestion
Traffic is expensive, especially in the cloud. Cloud providers charge different prices for traffic, depending on whether it is flowing within an availability zone, between availability zones (AZs), between particular regions, or to the open internet. Cross-regional and egress traffic is not only the most expensive, but also the most vulnerable to errors, latency, and security threats. 
```

> Network Performance Monitoring allows you to track all of the traffic patterns described above by mapping dependencies between any tags in Datadog, including service, container, availability zone, region, datacenter, etc… This insight into your dependencies and the amount of traffic volume they produce (which is ultimately what you are billed on) can be used to monitor and optimize on your traffic-related costs. 

Reviewer 2: ```suggestion
Network Performance Monitoring allows you to track all of the traffic patterns described above by mapping dependencies between any tags in Datadog, including service, container, availability zone, region, datacenter, etc… This insight into your dependencies and the amount of traffic volume they produce (which is ultimately what cloud providers charge for) can be used to monitor and optimize on your traffic-related costs. 
```

## Datadog’s story

> When we began our Kubernetes migration, migrating stateless services was (expectedly) much faster and easier than migrating our stateful services (e.g. Kafka), so we migrated our stateless services first. The outcome was terabytes of cross-AZ traffic between the stateful services (all in one AZ) and stateless services (spread across the other AZs), but then our cloud bill exploded. We used our own Network Product to identify the root cause - a suboptimal migration strategy - to optimize our network (by sharding our stateful services) and ultimately saw significant reductions in our cloud provider costs.

Reviewer 2: ```suggestion
When we began our Kubernetes migration, migrating stateless services was (expectedly) much faster and easier than migrating our stateful services (e.g. Kafka), so we migrated our stateless services first. The outcome was an terabytes of new cross-AZ traffic between the stateful services (all in one AZ) and stateless services (spread across the other AZs), which led our cloud bill to increase drastically and unexpectedly. We used our own NPM product to identify the root cause - a suboptimal migration strategy and consequently, inefficient and costly network communication. By sharding our stateful services, we were ultimately able to see significant (million-dollar) reductions in our cloud provider traffic costs.
```

## Steps for managing traffic costs 

> 1. To find similar issues in your own environment, you can start by grouping your traffic flows by regions, 

Reviewer 2: ```suggestion
1. To find similar issues in your own environment, you can start by scoping your view to traffic between regions, 
```

>     {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/region.png" alt="Group flows by region">}}

Reviewer 2: one screenshot is enough here to convey the point - let's keep region and remove the rest

    availability zones,
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/availability_zone.png" alt="Group flows by availability">}}
    and datacenters:
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/datacenter.png" alt="Group flows by datacenter">}}
    An increase in your traffic bill is almost always tied to an increase in one of these types of traffic. In many cases you may want to group traffic by asymmetric search terms - meaning, you want to see the source of the traffic in terms of one tag, but want to see the destination in terms of another. This kind of asymmetric query can be used to identify costly dependencies between your on-prem datacenters and cloud regions, 
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/aws_account.png" alt="Identify dependencies between datacenters and cloud regions">}}

>     as well as between clouds. One particularly helpful view is to group the source of the traffic by service, and the destination across availability zones. This will help you identify particularly noisy and expensive services that are communicating across multiple availability zones, like what we found during our migration.

Reviewer 2: a screenshot here would be helpful, with a caption

I'm thinking a query like: 
Source filter: AZ:1 | Source group by: service 
Destination filter: -AZ:1 | Destination group by: AZ

when sorting the table by volume, this would show the top X services communicating most from AZ:1 to other AZ's

>     as well as between clouds. One particularly helpful view is to group the source of the traffic by service, and the destination across availability zones. This will help you identify particularly noisy and expensive services that are communicating across multiple availability zones, like what we found during our migration.

Reviewer 2: The last sentence is confusing w/out a screenshot IMO. Since we provide more details about it in step 2, I'd remove it.

2. From here, isolate services that have the most traffic volume across multiple AZs. You can use filters inside the search bars to narrow down your query - so, for example, you can show only services that originate inside of one availability zone, and are sending traffic to any availability zone other than that one.
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/service_availability_zone.png" alt="Highlight any services that communicated across AZs">}}

>     The query above highlights any services that communicate across AZs, from us-east4-a to anywhere else. The chattiest services surfaced by this query may contribute heavily to overall cross-AZ traffic volume, and we can narrow it down even further by searching for one particular service, grouping the traffic by availability zone.

Reviewer 2: ```suggestion
    The query above highlights any services that communicate across AZs, from us-east4-a to anywhere else. Since the table is already sorted by volume, the first few rows surface the chattiest services contributing to the most cross-AZ traffic. If we want to inspect the cross-infrastructure effects of one of these culprits, we can filter the Source to that particular service and view its traffic to every other availability zone.
```

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/single_service.png" alt="Search for a single service.">}}

3. Similarly, you can use the team tag to identify engineering teams that generate the most, for instance, cross-regional traffic 
{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/team_region.png" alt="Use the team tag.">}}
or monitor your own team’s output specifically.
{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/region_region.png" alt="Use the region tag.">}}

4. To monitor costs from external traffic, scope your destination endpoints to public IPs using the IP Type facet.
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/scope_destination_points.png" alt="Use the type facet.">}}

>     Then group your destination by `dns` to break down external traffic volume by where it’s going. Although you cannot install an agent on public servers, we are able to resolve external endpoints to human-readable domain names using DNS resolution. 

Reviewer 2: ```suggestion
    Then group your destination by `domain` to break down external traffic volume by where it’s going. Although you cannot install an agent on public servers, we are able to resolve IPs representing external and cloud endpoints to human-readable domain names. 
```

Reviewer 2: we can link [external and cloud endpoints to human-readable domain names][https://www.datadoghq.com/blog/dns-resolution-datadog/]

    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/dns_resolution.png" alt="Group by DNS.">}}
    The example query above filters for traffic to AWS S3, elastic load balancers, APIs, and external ‘.com’ domains using substring wildcard entries (e.g. dns:*s3*).  
    {{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/wildcard.png" alt="Search with wildcards">}}

## Visualizing Traffic Costs 

> You can visualize cross-AZ or inter-AZ traffic using the Network Map to quickly pinpoint bottlenecks. At Datadog, we also use this view to validate that our EU and US availability zones are not communicating, for compliance reasons. 

Reviewer 2: ```suggestion
You can visualize cross-AZ or inter-AZ traffic using the Network Map to quickly pinpoint bottlenecks. At Datadog, we also use this view to validate that our EU and US availability zones are not communicating, to ensure compliance with GDPR and protection of our customers' data. 
```

(Cross-AZ traffic)
{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/cross-az-traffic.png" alt="Cross-AZ traffic">}}
(Inter-AZ service-to-service traffic) 
{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/inter-az-service-to-service-traffic.png" alt="Inter-AZ service-to-service traffic">}}

> By default, the thickness of the gray lines between the availability zones indicates the volume of traffic flowing between them, which is what contributes to your costs.

Reviewer 2: ```suggestion
Thick edges between the map's nodes, representing availability zones, indicates high volume of traffic flowing between them, which is what contributes to your costs.
```

You can edit your preferences using the ‘Filter traffic’ button. In larger environments, we recommend scoping to just the most significant traffic sources by moving the sliders to include just the highest-volume dependencies.

{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/filter-traffic.png" alt="Scope your traffic">}}

## Graphing Traffic Costs 

We recommend tracking traffic volume metrics over time in dashboards and notebooks. You can graph traffic between any two endpoints using the same queries you would make in the network page. Just create a Timeseries Widget and select the ‘Network Traffic’ source from the dropdown menu.  

{{< img src="network_performance_monitoring/guide/manage_traffic_costs_with_npm/timeseries.png" alt="Create a Timeseries">}}

> You can also use this data as the source of a monitor - so you can get a notification if traffic volume increases unexpectedly. This can give you a heads up that there’s something unusual happening before you get your cloud bill at the end of the month.

Reviewer 2: We don't have monitors yet! So let's remove this sentence, and replace with something about how dashboards and notebooks make it easy for you to share any issues you spot with your teammates.
